src/quic/node_quic_stream.cc

```diff
@@ -139,21 +193,18 @@ void QuicStream::Destroy() {
 int QuicStream::DoShutdown(ShutdownWrap* req_wrap) {
   if (is_destroyed())
     return UV_EPIPE;
-  Debug(this, "Shutdown writable side");
-  // Do nothing if the stream was already shutdown. Specifically,
-  // we should not attempt to send anything on the QuicSession
-  if (!is_writable())
-    return UV_EPIPE;
-  stream_stats_.closing_at = uv_hrtime();
-  set_write_close();
 
-  // If we're not currently within an ngtcp2 callback, then we need to
-  // tell the QuicSession to initiate serialization and sending of any
-  // pending frames.
-  if (!QuicSession::Ngtcp2CallbackScope::InNgtcp2CallbackScope(session_.get()))
-    session_->SendStreamData(this);
+  QuicSession::SendSessionScope send_scope(session());
+
+  if (is_writable()) {
+    Debug(this, "Shutdown writable side");
+    stream_stats_.closing_at = uv_hrtime();
+    set_write_close();
+    session()->ResumeStream(stream_id_);
+  }
 
-  return 1;
+  req_wrap->Done(0);
+  return 0;
```

Comment:
Why this? `return 1;` is exactly the right thing to indicate synchronous success:

https://github.com/nodejs/quic/blob/399cc9cbd9ec7cbe9805f5af45a35d097489a9a0/src/stream_base.h#L218-L221

Reply:
Arg... I keep mixing it up with DoWrite, which has return 0 for success.

Reply:
Inspired by this: https://github.com/nodejs/node/pull/31283

I think we can align the behaviour with `DoWrite()` here if you like, it’s just that we currently don’t have any `StreamBase`s that finish writes fully synchronously afaik (and if they do, it’s through `DoTryWrite()` rather than `DoWrite()`…)

Reply:
Eventually, but I plan to explore moving away from StreamBase entirely before too long to see if a model like Jeremiah's BOB streams would be more efficient. I'll leave it like this for now.